### ISSUE TITLE
Wait for client FIN during server-initiated socket close

### ISSUE THREAD

**Issue by halter73** [aspnet/AspNetCore#4703]
Sometimes clients in the LargeUpload tests observe empty responses (#3022). I believe this happens because Kestrel behaves badly when it initiates the FIN which happens for "Connection: close" or HTTP/1.0 requests. Kestrel should send a FIN, *wait for a client FIN with some timeout*, and then dispose the socket. Today, we skip the second step in both the Socket and libuv transport which could be the cause of clients not observing all the buffered response data.

I've opened #3036 to test this theory. If the lLargeUpload tests stabilize after this change, that will serve as evidence we should change this behavior.

**Comment by davidfowl** [comment]
@halter73 should we close this now?

**Comment by halter73** [comment]
Yes. In my experience, the OS will keep the socket in the TIME_WAIT state even after the process exits (let alone socket close) and retransmit as necessary. We can open a new issue if there's customer feedback that this is not the case.
